ocaml-lsp-server/src/state.ml

```diff
@@ -13,5 +13,6 @@ type t =
   ; configuration : Configuration.t
   ; trace : TraceValue.t
   ; ocamlformat : Fmt.t
+  ; diagnostics : Diagnostics.t
```

Comment:
Why is `Diagnostic.t` in both `State.t` and `Dune.t`?

Reply:
Good question. We do make sure to only mutate` Diagnostics.t` in the main thread, so that gives a lot of safety right away.
The only issue that could come up is with `Diagnostics.send` and other functions mutating the state. However, I still can't see anything that might go wrong since `send` doesn't wait on a fiber before mutating anything.

Reply:
I still find it a bit confusing that we have multiple instances of `diagnostics`. It make it hard to make a mental model of the diagnostics workflow:

- Where are the diagnostics sent from?
- Where are Dune diagnostics and merlin diagnostics stored, in which instance?

Could we pass a function to add/remove diagnostics to the modules that need access to it instead? It makes sense to have a global instance of Diagnostics in the state IMO, so that the diagnostics IO can be centralized.

Reply:
> I still find it a bit confusing that we have multiple instances of diagnostics. It make it hard to make a mental model of the diagnostics workflow:

We do not have multiple instances of `Diagnostics.t` - `Diagnostics.create` is only called once. We have two references to it in `State.t` and `Dune.t`. That's mostly because `Dune.t` cannot refer to `State.t` due to cycles.

> Could we pass a function to add/remove diagnostics to the modules that need access to it instead? It makes sense to have a global instance of Diagnostics in the state IMO, so that the diagnostics IO can be centralized.

Maybe you could post a diff to show what you mean? I don't quite see where this is going. In my mind, the current code already achieves  the goal of centralizing the sending of diagnostics: there's a `Diagnostics.t` value and it is responsible for sending diagnostics